src/day2/scan_passwds.rb

```diff
@@ -0,0 +1,100 @@
+#! /usr/bin/env -S ruby -w
+
+#: AoC Day 2 - Count valid passwords. (See `problems.txt` for the exact
+#: problems definitions.)
+
+require 'optparse'
+require 'ostruct'
+
+module ScanPasswds
+
+  extend self
+
+  #: Return true if given passwd matches original shopkeeper's policies
+  #: constraints, false otherwise. (See day2/problems.txt for details.)
+  def match_constraints_orig?(min, max, char, passwd)
+    count = passwd.count char
+    min <= count && count <= max
+  end
+
+  #: Return true if given passwd matches Toboggan Corporate Policies (TCP)
+  #: constraints, false otherwise. (Note that TCP policy authors count from 1.
+  #: See day2/problems.txt for details.)
+  def match_constraints_tcp?(first, second, char, passwd)
+    (passwd[first - 1] == char) ^ (passwd[second - 1] == char)
```

Comment:
What does this `^` do in this case?

Reply:
Xor. This is an implementation of the "Toboggan Corporate Policy" for authentication, namely:
> Exactly one of these positions must contain the given letter.

Thus `match_constraints_tcp?` returns `true` iff exactly one of the positions contains the required character.

Reply:
OK, cool. Now I learned this operator can be used for booleans as well as bit math.